### PR TITLE
Make various types public as they were already visible externally

### PR DESCRIPTION
**Pull request by adityapatwardhan**
## PR Summary

Many types were already visible publicly either as attributes or as OutputType. This change marks them as public.

- `ValidateVariableNameAttribute`  - is used by a public property `CommonParameters.ErrorVariable`
- `ArgumentToVersionTransformationAttribute` - is used by a public property `SetStrictModeCommand.Version`
- `ValidateVersionAttribute` - is used by a public property `SetStrictModeCommand.Version`
- `EnumSingleTypeConverter` and `EnumMultipleTypeConverter ` - classes had public methods
- `ArgumentToModuleTransformationAttribute` - used by `Save-Help` command for `Module` parameter.
- `ArgumentToEncodingTransformationAttribute` - used by multiple cmdlets for `Encoding` parameter.

Also, changed the output type of `Start-Job` from `PSRemotingJob` to `Job`

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.  
- **User-facing changes**
    - [x] Not Applicable
    - **OR**  
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [x] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [ ] [Add `[feature]` to your commit messages if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

### CODE REVIEW

**Review by iSazonov — COMMENTED** [review]
For new public APIs we need PowerShell Committee approval.

**Comment by PaulHigin** [comment]
What is the motivation for making these types public?  For example, I believe `PSRemotingJob` was meant for internal implementation and only its base class `Job` was meant for public use.

File: src/System.Management.Automation/engine/ArgumentTypeConverterAttribute.cs [hunk 1]
```diff
@@ -36,6 +36,9 @@ internal Type TargetType
             }
         }
 
+        /// <summary>
+        /// Transforms the inputData
```
[the remark this answers is not in the document]
Comment (adityapatwardhan) [reply src/System.Management.Automation/engine/ArgumentTypeConverterAttribute.cs]: Fixed.

File: src/System.Management.Automation/utils/EncodingUtils.cs [hunk 2]
```diff
@@ -87,8 +87,14 @@ internal static Encoding Convert(Cmdlet cmdlet, string encoding)
...
     /// </summary>
-    internal sealed class ArgumentToEncodingTransformationAttribute : ArgumentTransformationAttribute
+    public sealed class ArgumentToEncodingTransformationAttribute : ArgumentTransformationAttribute
     {
+        /// <summary>
+        /// Transform the argument to type System.Text.Encoding.
+        /// <param name="engineIntrinsics">EngineIntrinsics object.</param>
+        /// <param name="inputData">Value of attribute to be transformmed.</param>
+        /// <returns>The object after transformation.</returns>
+        /// </summary>
```
[the remark this answers is not in the document]
Comment (adityapatwardhan) [reply src/System.Management.Automation/utils/EncodingUtils.cs]: Fixed.

File: src/System.Management.Automation/engine/CommonCommandParameters.cs [hunk 3]
```diff
@@ -230,8 +230,20 @@ public string PipelineVariable
...
+            /// <summary>
+            /// Validate the value of arguments is a valid variable name.
+            /// <param name="arguments">
+            /// The value of the arguments is a variable name to be validated.
+            /// </param>
+            /// <param name="EngineIntrinsics">
+            /// Instance of <see cref="EngineIntrinsics"> class.
+            /// </param>
+            /// </summary>
             protected override void Validate(object arguments, EngineIntrinsics engineIntrinsics)
```
[the remark this answers is not in the document]
Comment (adityapatwardhan) [reply src/System.Management.Automation/engine/CommonCommandParameters.cs]: I do not want to make any changes unless needed.

File: src/System.Management.Automation/engine/InternalCommands.cs [hunk 4]
```diff
@@ -2043,8 +2050,16 @@ public override object Transform(EngineIntrinsics engineIntrinsics, object input
...
+        /// Attribute to validate the version.
+        /// </summary>
+        public sealed class ValidateVersionAttribute : ValidateArgumentsAttribute
         {
+            /// <summary>
+            /// Validate the arguments is a valid version.
+            /// <param name="arguments">value of the version argument.</param>
+            /// <param name="engineIntrinsics">EngineIntrinsics object.</param>
+            /// </summary>
             protected override void Validate(object arguments, EngineIntrinsics engineIntrinsics)
```
[the remark this answers is not in the document]
Comment (adityapatwardhan) [reply src/System.Management.Automation/engine/InternalCommands.cs]: I do not want to make any changes unless needed.

**Review by JamesWTruher — COMMENTED** [review]
I don't enough context to suggest that there are any problems with this

**Comment by lzybkr** [comment]
I'm not totally sure I understand the motivation - maybe an upcoming change in the CLR?

I haven't reviewed the changes closely, but after a quick glance, if a CLR change forces these types to be public, I'd move them to an `Internal` namespace to make it clear they are implementation details, and probably add an `Obsolete` attribute to discourage/disallow their use in other assemblies.

**Comment by adityapatwardhan** [comment]
@lzybkr - We are planning to use [GenAPI](https://github.com/dotnet/buildtools) to build reference assemblies. These types are not found when compiling the generated `.cs` file, as they are marked as internal. I believe all of them should be marked as public, hence the change.

**Comment by lzybkr** [comment]
I see. I'm assuming that custom attributes are always included in the reference assembly just in case a build tool like the C# compiler needs to see the attribute to generate correct code or check for errors.

I'm not aware of any build tool that needs to see PowerShell's attributes, so I would expect careful thought before blindly making these types public.

Many of the names are not good public apis - they maybe sound too general, e.g.
* `ArgumentToVersionTransformationAttribute` is quite specific and special w.r.t. PowerShell versions
* `ValidateVariableName` is missing the `Attribute` suffix and validates in a very specific way that is not as general as the name suggests.
* `ArgumentToModuleTransformationAttribute` sounds promising and useful, yet it's only used in one place in PowerShell - so maybe it's not really that useful.

I wonder if `EnumMultipleTypeConverter` needs to be applied as an attribute at all - it's applied on types that are also `Flags`, so it's sort of ambiguous - should the conversion produce a single value or multiple values. If that could be removed, then maybe you can avoid exposing the `Enum*TypeConverter` types.

**Comment by daxian-dbw** [comment]
> EnumSingleTypeConverter and EnumMultipleTypeConverter - classes had public methods

Having public methods doesn't seem to be a strong argument to make them public. We have other internal types with `public` modifier applied to its methods or properties. These 2 types are not attributes, so how are they exposed by `GenApi`?

**Comment by iSazonov** [comment]
If GenAPI doesn't work because the types is visible externally it is GenAPI problem and it would be fixed in GenAPI. Or we could create a workaround on PowerShell to resolve the problem.

**Comment by adityapatwardhan** [comment]
I will shortly add a gist of the `.cs` file generated by `GenAPI`.

**Pull request closed without merging**